File: src/csc_presor.f
!                    *********************
                     SUBROUTINE CSC_PRESOR
!                    *********************
     & (MA,MP,MQ,W,NAMP,NAMQ)
!
!***********************************************************************
! CSC PACKAGE - SERGIO CASTIBLANCO
!***********************************************************************
!
!brief    1) PREPARES THE MATRICES FOR SUCCESIVE OVER-RELAXATION SOLVER
!
!history  Sergio Castiblanco
!+        25/01/2021
!+        Translation for original Matlab implementation
!
!+
!!
!! P and Q:
!! Classic iterative methods use the next recurrence equation to solve 
!! the linear system of equations Ax=b:
!! 
!!               P * x(t+1) = Q * x(t) + b
!!
!!       For SOR:
!!            P = ((1/w)*D + L)
!!            Q = ((1/w - 1)*D - U); where: L = lower part of A
!!                                          U = upper part of A
!!                                          D = diagonal of A
!!
!! For the particular case when w = 1, we are in Gauss-Seidel method
!!
!!
!!      Sergio A. Castiblanco B. - Advanced Numerical Methods 
!!      Pontificia Universidad Javeriana - Bogota
!!
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!| MA       |-->| MATRIX A IN CSC                                      |
!| MP       |<->| MATRIX P IN CSC                                      |
!| MQ       |<->| MATRIX Q IN CSC                                      |
!| W        |<->| OVER-RELAXATION COEFFICIENT                          |
!| NAMP     |-->| NAME OF MATRIX P                                     |
!| NAMQ     |-->| NAME OF MATRIX Q                                     |
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      USE CSC_STORAGE, EX_CSC_PRESOR => CSC_PRESOR
      IMPLICIT NONE
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      TYPE(CSC_OBJ), INTENT(IN) :: MA
      TYPE(CSC_OBJ), INTENT(INOUT) :: MP,MQ
      DOUBLE PRECISION, INTENT(IN) :: W
      CHARACTER(LEN=6), INTENT(IN) :: NAMP,NAMQ
!
!  IN SUBROUTINE VARIABLES
!
      INTEGER :: J, I, QP, PP
      INTEGER :: NZP, NZQ
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
! COMPUTING NZP AND NZQ
!
      NZP = 0
      NZQ = 0
!
      DO J = 1,MA%NC
        DO I = MA%C(J),MA%C(J+1)-1
          IF(MA%R(I).LT.J) THEN
            NZQ = NZQ + 1
          ELSEIF(MA%R(I).GT.J) THEN
            NZP = NZP + 1
          ELSEIF(MA%R(I).EQ.J) THEN
            IF ((W.GT.(1.D0-1E-8)).AND.(W.LT.1.D0+1E-8)) THEN
              NZP = NZP + 1
            ELSE
              NZP = NZP + 1
              NZQ = NZQ + 1
            ENDIF
          ENDIF
        ENDDO
      ENDDO
! 
! ALLOCATING MP AND MQ
!
      CALL ALL_CSC(MP, MA%NR, MA%NC, NZP, NAMP)
      CALL ALL_CSC(MQ, MA%NR, MA%NC, NZQ, NAMQ)
!
! FILLING MP AND MQ
!
      MQ%C(1) = 1
      MP%C(1) = 1
!
      QP = 1
      PP = 1
      IF((W.GT.(1.D0-1E-8)).AND.(W.LT.1.D0+1E-8)) THEN
        DO J = 1,MA%NC
          DO I = MA%C(J),MA%C(J+1)-1
            IF(MA%R(I).LT.J) THEN
              MQ%V(QP) = -MA%V(I)
              MQ%R(QP) = MA%R(I)
              QP = QP + 1
            ELSE
              MP%V(PP) = MA%V(I)
              MP%R(PP) = MA%R(I) 
              PP = PP + 1
            ENDIF
          ENDDO
          MQ%C(J+1) = QP
          MP%C(J+1) = PP
        ENDDO
      ELSE
        DO J = 1,MA%NC
          DO I = MA%C(J),MA%C(J+1)-1
            IF(MA%R(I).LT.J) THEN
              MQ%V(QP) = -MA%V(I)
              MQ%R(QP) = MA%R(I)
              QP = QP + 1
            ELSEIF(MA%R(I).GT.J) THEN
              MP%V(PP) = MA%V(I)
              MP%R(PP) = MA%R(I) 
              PP = PP + 1
            ELSE
              MQ%V(QP) = ((1D0/W) - 1D0)*MA%V(I)
              MQ%R(QP) = MA%R(I)
              QP = QP + 1
              MP%V(PP) = (1D0/W)*MA%V(I)
              MP%R(PP) = MA%R(I)
              PP = PP + 1
            ENDIF
          ENDDO
          MQ%C(J+1) = QP
          MP%C(J+1) = PP
        ENDDO
      ENDIF
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      END SUBROUTINE CSC_PRESOR

File: matlab/csc_preSOR.m
function [Pv,Pr,Pc,Qv,Qr,Qc] = csc_preSOR(Av,Ar,Ac,w)
%
% This function takes the matrix A in CSC storage, and prepares the
% matrices for solve the system Ax=b with SOR iterative method
%
% P and Q:
% Classic iterative methods use the next recurrence equation for solve the 
% system:
% 
%               P * x(t+1) = Q * x(t) + b
%
%       For SOR:
%            P = ((1/w)*D + L)
%            Q = ((1/w - 1)*D - U); where: L = lower part of A
%                                          U = upper part of A
%                                          D = diagonal of A
%
% For the particular case when w = 1, we are in Gauss-Seidel method
%
% Entries:
%     Av, Ar, Ac : Components of matrix A in CSC stotage
%     w : SOR Over-Relaxation coefficient
%
%      Sergio A. Castiblanco B. - Métodos Numéricos Avanzados
%      Pontificia Universidad Javeriana - Bogotá
%

m = length(Ac)-1;
nzp = 0;
nzq = 0;

for j=1:m
    for i=Ac(j):Ac(j+1)-1
        if Ar(i)<j
            nzq = nzq + 1;
        elseif Ar(i)>j
            nzp = nzp + 1;
        elseif Ar(i)==j
            if w==1
                nzp = nzp+1;
            else
                nzp = nzp+1;
                nzq = nzq+1;
            end
        end
    end
end

Qv = zeros(nzq,1);
Qr = zeros(nzq,1);
Qc = zeros(m+1,1);
Qc(1) = 1;
Pv = zeros(nzp,1);
Pr = zeros(nzp,1);
Pc = zeros(m+1,1);
Pc(1) = 1;
    
qp=1;
pp=1;
if w==1
    for j=1:m
        for i=Ac(j):Ac(j+1)-1
            if Ar(i) < j
                Qv(qp) = -Av(i);
                Qr(qp) = Ar(i);
                qp = qp + 1;
            else
                Pv(pp) = Av(i);
                Pr(pp) = Ar(i);
                pp = pp + 1;
            end
        end
        Qc(j+1) = qp;
        Pc(j+1) = pp;
    end
else
    for j=1:m
        for i=Ac(j):Ac(j+1)-1
            if Ar(i) < j
                Qv(qp) = -Av(i);
                Qr(qp) = Ar(i);
                qp = qp + 1;
            elseif Ar(i) > j
                Pv(pp) = Av(i);
                Pr(pp) = Ar(i);
                pp = pp + 1;
            else
                Qv(qp) = (1/w - 1)*Av(i);
                Qr(qp) = Ar(i);
                qp = qp + 1;
                Pv(pp) = (1/w)*Av(i);
                Pr(pp) = Ar(i);
                pp = pp + 1;
            end
        end
        Qc(j+1) = qp;
        Pc(j+1) = pp;
    end
end

end